central/platform/reprocessor/reprocessor_impl.go

```diff
@@ -0,0 +1,170 @@
+	"sync/atomic"
+
+	alertDS "github.com/stackrox/rox/central/alert/datastore"
+	alertutils "github.com/stackrox/rox/central/alert/utils"
+	deploymentDS "github.com/stackrox/rox/central/deployment/datastore"
+	platformmatcher "github.com/stackrox/rox/central/platform/matcher"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/search"
+)
+
+const batchSize = 5000
+
+var (
+	log = logging.LoggerForModule()
+
+	reprocessorCtx = sac.WithAllAccess(context.Background())
+)
+
+type platformReprocessorImpl struct {
+	alertDatastore      alertDS.DataStore
+	deploymentDatastore deploymentDS.DataStore
+	platformMatcher     platformmatcher.PlatformMatcher
+
+	// isStarted will make sure only one reprocessing routine runs for an instance of reprocessor
+	isStarted atomic.Bool
+}
+
+func New(alertDatastore alertDS.DataStore,
+	deploymentDatastore deploymentDS.DataStore,
+	platformMatcher platformmatcher.PlatformMatcher) PlatformReprocessor {
+
+	return &platformReprocessorImpl{
+		alertDatastore:      alertDatastore,
+		deploymentDatastore: deploymentDatastore,
+		platformMatcher:     platformMatcher,
+	}
+}
+
+func (pr *platformReprocessorImpl) Start() {
+	swapped := pr.isStarted.CompareAndSwap(false, true)
+	if !swapped {
+		log.Error("Platform reprocessor already running")
+		return
+	}
+	go pr.runReprocessing()
+}
+
+func (pr *platformReprocessorImpl) Stop() {
+	if !pr.isStarted.Load() {
+		log.Error("Platform reprocessor not started")
+		return
+	}
+}
+
+func (pr *platformReprocessorImpl) runReprocessing() {
+	needsReprocessing, err := pr.needsReprocessing()
+	if err != nil {
+		log.Errorf("Error determining if platform components need reporcessing: %v", err)
+		return
+	}
+
+	if !needsReprocessing {
+		log.Info("Platform components up to date, skipping reprocessing")
+		return
+	}
+
+	err = pr.reprocessAlerts()
+	if err != nil {
+		log.Errorf("Error reprocessing alerts with platform rules: %v", err)
+		return
+	}
+
+	err = pr.reprocessDeployments()
+	if err != nil {
+		log.Errorf("Error reprocessing deployments with platform rules: %v", err)
+		return
+	}
+}
+
+func (pr *platformReprocessorImpl) needsReprocessing() (bool, error) {
+	query := search.NewQueryBuilder().AddNullField(search.PlatformComponent).ProtoQuery()
+	numAlerts, err := pr.alertDatastore.Count(reprocessorCtx, query)
+	if err != nil {
+		return false, err
+	}
+	numDeployments, err := pr.deploymentDatastore.Count(reprocessorCtx, query)
+	if err != nil {
+		return false, err
+	}
+	return numAlerts > 0 || numDeployments > 0, nil
+}
+
+func (pr *platformReprocessorImpl) reprocessAlerts() error {
+	query := search.EmptyQuery()
+	query.Pagination = &v1.QueryPagination{
+		Limit:  batchSize,
+		Offset: 0,
+		SortOptions: []*v1.QuerySortOption{
+			{
+				Field: search.AlertID.String(),
+			},
+		},
+	}
+
+	for {
+		alerts, err := pr.alertDatastore.SearchRawAlerts(reprocessorCtx, query)
```

Comment:
This function appends a default `OR` condition on all queries for "reasons" (IMO that should have been done in the service layer).  In fact most the search functions of alerts do.  That is why I had to use `WalkByQuery` for pruning.  I think you will have to do something similar or miss tagging alerts and thus getting stuck ALWAYS reprocessing because you never see some of the NULL rows when re-tagging.

https://github.com/stackrox/stackrox/blob/146467b3c74ec069cae27aa8796dafbd30cd8697/central/alert/datastore/internal/search/searcher_impl.go#L124

Reply:
Looks like it adds an `AND` condition to the query, not an `OR` . So the `PlatformComponent`  field would always stay `Null`  for resolved and snoozed alerts before upgrade. That's an even bigger issue than the task running on startup every time.

Reply:
Since `applyDefaultState` only changes the query if it does not explicitly specify a `Violation State` filter, I updated the query to explicitly add the filter such that it will match all possible violation states. Also created a ticket to move out `applyDefaultState` from the searcher.

Reply:
This seems hacky and not maintainable to me.  Would it not be simpler to use `WalkByQuery` OR to simply add `GetByQuery` to the datastore and use it?  `GetByQuery` exists in the store so should just be a matter of exposing it in the datastore.

Reply:
Well I did not think of `GetByQuery`. Yes that can work. 
I did not want to use `WalkByQuery` to keep the `needsReprocessing` implementation simpler and just get a count.